File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/bytecode/ByteCodeUtils.java
package net.yakclient.mixin.base.internal.bytecode;

import net.yakclient.mixin.mixin.base.internal.ASMType;
import org.jetbrains.annotations.Nullable;
import org.objectweb.asm.Opcodes;

import java.io.DataInputStream;
import java.io.IOException;

public class ByteCodeUtils {
    public static final ASMType DEFAULT_ASM_MODE = ASMType.TREE;
    public static final int ASM_VERSION = Opcodes.ASM9;


    public static String removeReturnType(String type) {
        return type.substring(0, type.indexOf(')') + 1);
    }

    public static boolean descriptorsSame(String method1, String method2) {
        return removeReturnType(method1).equals(removeReturnType(method2));
    }

    public static char primitiveType(Class<?> cls) {
        if (cls == void.class) return 'V';
        if (cls == boolean.class) return 'Z';
        if (cls == char.class) return 'C';
        if (cls == byte.class) return 'B';
        if (cls == short.class) return 'S';
        if (cls == int.class) return 'I';
        if (cls == float.class) return 'F';
        if (cls == long.class) return 'J';
        if (cls == double.class) return 'D';
        throw new IllegalArgumentException("Class must be a primitive!");
    }

    public static Class<?> primitiveType(char c) {
        if (c == 'V') return void.class;
        if (c == 'Z') return boolean.class;
        if (c == 'C') return char.class;
        if (c == 'B') return boolean.class;
        if (c == 'S') return short.class;
        if (c == 'I') return int.class;
        if (c == 'F') return float.class;
        if (c == 'J') return long.class;
        if (c == 'D') return double.class;
        throw new IllegalArgumentException("Class must be a primitive!");
    }

    public static boolean isPrimitiveType(char c) {
        switch (c) {
            case 'V':
            case 'Z':
            case 'C':
            case 'B':
            case 'S':
            case 'I':
            case 'F':
            case 'J':
            case 'D':
                return true;
        }
        return false;
    }

    public static boolean classExists(String clazz) {
        try {
            return loadClassBytes(clazz) != null;
        } catch (IOException | ClassNotFoundException e) {
            return false;
        }
    }

    @Nullable
    public static byte[] loadClassBytes(String name) throws IOException, ClassNotFoundException {
        try (var cIn = ClassLoader.getSystemResourceAsStream(name.replace('.', '/') + ".class")) {
            if (cIn == null) return null;
            try (var in = new DataInputStream(cIn)) {
                final var buf = new byte[cIn.available()];

                in.readFully(buf);
                return buf;
            }
        }
    }

    /**
     * This is purely meant for debugging/testing purposes, it will be slow
     * and has no use during production runtime.
     *
     * @param opcode The opcode to translate.
     * @return The name of the opcode.
     * @throws IllegalArgumentException If the opcode is not found.
     */
    public static String opcodeToString(int opcode) throws IllegalArgumentException {
        switch (opcode) {
            case 0:
                return "NOP";
            case 1:
                return "ACONST_NULL";
            case 2:
                return "ICONST_M1";
            case 3:
                return "ICONST_0";
            case 4:
                return "ICONST_1";
            case 5:
                return "ICONST_2";
            case 6:
                return "ICONST_3";
            case 7:
                return "ICONST_4";
            case 8:
                return "ICONST_5";
            case 9:
                return "LCONST_0";
            case 10:
                return "LCONST_1";
            case 11:
                return "FCONST_0";
            case 12:
                return "FCONST_1";
            case 13:
                return "FCONST_2";
            case 14:
                return "DCONST_0";
            case 15:
                return "DCONST_1";
            case 16:
                return "BIPUSH";
            case 17:
                return "SIPUSH";
            case 18:
                return "LDC";
            case 21:
                return "ILOAD";
            case 22:
                return "LLOAD";
            case 23:
                return "FLOAD";
            case 24:
                return "DLOAD";
            case 25:
                return "ALOAD";
            case 46:
                return "IALOAD";
            case 47:
                return "LALOAD";
            case 48:
                return "FALOAD";
            case 49:
                return "DALOAD";
            case 50:
                return "AALOAD";
            case 51:
                return "AALOAD";
            case 52:
                return "CALOAD";
            case 53:
                return "SALOAD";
            case 54:
                return "ISTORE";
            case 55:
                return "LSTORE";
            case 56:
                return "FSTORE";
            case 57:
                return "DSTORE";
            case 58:
                return "ASTORE";
            case 79:
                return "IASTORE";
            case 80:
                return "LASTORE";
            case 81:
                return "FASTORE";
            case 82:
                return "DASTORE";
            case 83:
                return "AASTORE";
            case 84:
                return "BASTORE";
            case 85:
                return "CASTORE";
            case 86:
                return "SASTORE";
            case 87:
                return "POP";
            case 88:
                return "POP2";
            case 89:
                return "DUP";
            case 90:
                return "DUP_X1";
            case 91:
                return "DUP_X2";
            case 92:
                return "DUP2";
            case 93:
                return "DUP2_X1";
            case 94:
                return "DUP2_X2";
            case 95:
                return "SWAP";
            case 96:
                return "IADD";
            case 97:
                return "LADD";
            case 98:
                return "FADD";
            case 99:
                return "DADD";
            case 100:
                return "ISUB";
            case 101:
                return "LSUB";
            case 102:
                return "FSUB";
            case 103:
                return "DSUB";
            case 104:
                return "IMUL";
            case 105:
                return "LMUL";
            case 106:
                return "FMUL";
            case 107:
                return "DMUL";
            case 108:
                return "IDIV";
            case 109:
                return "LDIV";
            case 110:
                return "FDIV";
            case 111:
                return "DDIV";
            case 112:
                return "IREM";
            case 113:
                return "LREM";
            case 114:
                return "FREM";
            case 115:
                return "DREM";
            case 116:
                return "INEG";
            case 117:
                return "LNEG";
            case 118:
                return "FNEG";
            case 119:
                return "DNEG";
            case 120:
                return "ISHL";
            case 121:
                return "LSHL";
            case 122:
                return "ISHR";
            case 123:
                return "LSHR";
            case 124:
                return "IUSHR";
            case 125:
                return "LUSHR";
            case 126:
                return "IAND";
            case 127:
                return "LAND";
            case 128:
                return "IOR";
            case 129:
                return "LOR";
            case 130:
                return "IXOR";
            case 131:
                return "LXOR";
            case 132:
                return "IINC";
            case 133:
                return "I2L";
            case 134:
                return "I2F";
            case 135:
                return "I2D";
            case 136:
                return "L2I";
            case 137:
                return "L2F";
            case 138:
                return "L2D";
            case 139:
                return "F2I";
            case 140:
                return "F2L";
            case 141:
                return "F2D";
            case 142:
                return "D2I";
            case 143:
                return "D2L";
            case 144:
                return "D2F";
            case 145:
                return "I2B";
            case 146:
                return "I2C";
            case 147:
                return "I2S";
            case 148:
                return "LCMP";
            case 149:
                return "FCMPL";
            case 150:
                return "FCMPG";
            case 151:
                return "DCMPL";
            case 152:
                return "DCMPG";
            case 153:
                return "IFEQ";
            case 154:
                return "IFNE";
            case 155:
                return "IFLT";
            case 156:
                return "IFGE";
            case 157:
                return "IFGT";
            case 158:
                return "IFLE";
            case 159:
                return "IF_ICMPEQ";
            case 160:
                return "IF_ICMPNE";
            case 161:
                return "IF_ICMPLT";
            case 162:
                return "IF_ICMPGE";
            case 163:
                return "IF_ICMPGT";
            case 164:
                return "IF_ICMPLE";
            case 165:
                return "IF_ACMPEQ";
            case 166:
                return "IF_ACMPNE";
            case 167:
                return "GOTO";
            case 168:
                return "JSR";
            case 169:
                return "RET";
            case 170:
                return "TABLESWITCH";
            case 171:
                return "LOOKUPSWITCH";
            case 172:
                return "IRETURN";
            case 173:
                return "LRETURN";
            case 174:
                return "FRETURN";
            case 175:
                return "DRETURN";
            case 176:
                return "ARETURN";
            case 177:
                return "RETURN";
            case 178:
                return "GETSTATIC ";
            case 179:
                return "PUTSTATIC";
            case 180:
                return "GETFIELD";
            case 181:
                return "PUTFIELD";
            case 182:
                return "INVOKEVIRTUAL ";
            case 183:
                return "INVOKESPECIAL";
            case 184:
                return "INVOKESTATIC";
            case 185:
                return "INVOKEINTERFACE";
            case 186:
                return "INVOKEDYNAMIC";
            case 187:
                return "NEW";
            case 188:
                return "NEWARRAY";
            case 189:
                return "ANEWARRAY";
            case 190:
                return "ARRAYLENGTH";
            case 191:
                return "ATHROW";
            case 192:
                return "CHECKCAST";
            case 193:
                return "INSTANCEOF";
            case 194:
                return "MONITORENTER";
            case 195:
                return "MONITOREXIT";
            case 197:
                return "MULTIANEWARRAY";
            case 198:
                return "IFNULL";
            case 199:
                return "IFNONNULL";
            default:
                throw new IllegalArgumentException("Invalid opcode");
        }
    }

    public static String bytesToHex(byte[] bytes) {
        final char[] HEX_ARRAY = "0123456789ABCDEF".toCharArray();

        final char[] hexChars = new char[bytes.length * 2];
        for (var j = 0; j < bytes.length; j++) {
            final int v = bytes[j] & 0xFF;
            hexChars[j * 2] = HEX_ARRAY[v >>> 4];
            hexChars[j * 2 + 1] = HEX_ARRAY[v & 0x0F];
        }
        return new String(hexChars);
    }

    public static boolean isReturn(int opcode) {
        switch (opcode) {
            case Opcodes.IRETURN:
            case Opcodes.LRETURN:
            case Opcodes.FRETURN:
            case Opcodes.DRETURN:
            case Opcodes.ARETURN:
            case Opcodes.RETURN:
                return true;
        }
        return false;
    }

    public static boolean isLocalsStore(int opcode) {
        switch (opcode) {
            case Opcodes.ASTORE:
            case Opcodes.DSTORE:
            case Opcodes.FSTORE:
            case Opcodes.ISTORE:
            case Opcodes.LSTORE:
                return true;
        }
        return false;
    }
}

File: mixin-base/src/main/kotlin/net/yakclient/mixin/base/internal/bytecode/ByteCodeUtils.kt
package net.yakclient.mixin.base.internal.bytecode

import net.yakclient.mixin.base.internal.ASMType
import net.yakclient.mixin.base.internal.instruction.Instruction
import org.jetbrains.annotations.TestOnly
import org.objectweb.asm.Opcodes
import org.objectweb.asm.tree.AbstractInsnNode
import org.objectweb.asm.tree.InsnList
import java.io.DataInputStream
import java.io.IOException
import java.util.function.Consumer

object ByteCodeUtils {
    val DEFAULT_ASM_MODE: ASMType = ASMType.TREE
    const val ASM_VERSION = Opcodes.ASM9


    fun removeReturnType(type: String?): String {
        return type!!.substring(0, type.indexOf(')') + 1)
    }

    @JvmStatic
    fun descriptorsSame(method1: String?, method2: String?): Boolean {
        return removeReturnType(method1) == removeReturnType(method2)
    }

    @JvmStatic
    fun primitiveType(cls: Class<*>): Char {
        if (cls == Void.TYPE) return 'V'
        if (cls == Boolean::class.javaPrimitiveType) return 'Z'
        if (cls == Char::class.javaPrimitiveType) return 'C'
        if (cls == Byte::class.javaPrimitiveType) return 'B'
        if (cls == Short::class.javaPrimitiveType) return 'S'
        if (cls == Int::class.javaPrimitiveType) return 'I'
        if (cls == Float::class.javaPrimitiveType) return 'F'
        if (cls == Long::class.javaPrimitiveType) return 'J'
        if (cls == Double::class.javaPrimitiveType) return 'D'
        throw IllegalArgumentException("Class must be a primitive!")
    }

    fun primitiveType(c: Char): Class<*>? {
        if (c == 'V') return Void.TYPE
        if (c == 'Z') return Boolean::class.javaPrimitiveType
        if (c == 'C') return Char::class.javaPrimitiveType
        if (c == 'B') return Boolean::class.javaPrimitiveType
        if (c == 'S') return Short::class.javaPrimitiveType
        if (c == 'I') return Int::class.javaPrimitiveType
        if (c == 'F') return Float::class.javaPrimitiveType
        if (c == 'J') return Long::class.javaPrimitiveType
        if (c == 'D') return Double::class.javaPrimitiveType
        throw IllegalArgumentException("Class must be a primitive!")
    }

    fun isPrimitiveType(c: Char): Boolean {
        when (c) {
            'V', 'Z', 'C', 'B', 'S', 'I', 'F', 'J', 'D' -> return true
        }
        return false
    }

    @JvmStatic
    fun classExists(clazz: String): Boolean {
        return try {
            loadClassBytes(clazz) != null
        } catch (e: IOException) {
            false
        } catch (e: ClassNotFoundException) {
            false
        }
    }

    @JvmStatic
    @Throws(IOException::class, ClassNotFoundException::class)
    fun loadClassBytes(name: String): ByteArray? {
        ClassLoader.getSystemResourceAsStream(name.replace('.', '/') + ".class").use { cIn ->
            if (cIn == null) return null
            DataInputStream(cIn).use { `in` ->
                val buf = ByteArray(cIn.available())
                `in`.readFully(buf)
                return buf
            }
        }
    }

    /**
     * This is purely meant for debugging/testing purposes, it will be slow
     * and has no use during production runtime.
     *
     * @param opcode The opcode to translate.
     * @return The name of the opcode.
     * @throws IllegalArgumentException If the opcode is not found.
     */
    @Throws(IllegalArgumentException::class)
    fun opcodeToString(opcode: Int): String {
        return when (opcode) {
            0 -> "NOP"
            1 -> "ACONST_NULL"
            2 -> "ICONST_M1"
            3 -> "ICONST_0"
            4 -> "ICONST_1"
            5 -> "ICONST_2"
            6 -> "ICONST_3"
            7 -> "ICONST_4"
            8 -> "ICONST_5"
            9 -> "LCONST_0"
            10 -> "LCONST_1"
            11 -> "FCONST_0"
            12 -> "FCONST_1"
            13 -> "FCONST_2"
            14 -> "DCONST_0"
            15 -> "DCONST_1"
            16 -> "BIPUSH"
            17 -> "SIPUSH"
            18 -> "LDC"
            21 -> "ILOAD"
            22 -> "LLOAD"
            23 -> "FLOAD"
            24 -> "DLOAD"
            25 -> "ALOAD"
            46 -> "IALOAD"
            47 -> "LALOAD"
            48 -> "FALOAD"
            49 -> "DALOAD"
            50 -> "AALOAD"
            51 -> "AALOAD"
            52 -> "CALOAD"
            53 -> "SALOAD"
            54 -> "ISTORE"
            55 -> "LSTORE"
            56 -> "FSTORE"
            57 -> "DSTORE"
            58 -> "ASTORE"
            79 -> "IASTORE"
            80 -> "LASTORE"
            81 -> "FASTORE"
            82 -> "DASTORE"
            83 -> "AASTORE"
            84 -> "BASTORE"
            85 -> "CASTORE"
            86 -> "SASTORE"
            87 -> "POP"
            88 -> "POP2"
            89 -> "DUP"
            90 -> "DUP_X1"
            91 -> "DUP_X2"
            92 -> "DUP2"
            93 -> "DUP2_X1"
            94 -> "DUP2_X2"
            95 -> "SWAP"
            96 -> "IADD"
            97 -> "LADD"
            98 -> "FADD"
            99 -> "DADD"
            100 -> "ISUB"
            101 -> "LSUB"
            102 -> "FSUB"
            103 -> "DSUB"
            104 -> "IMUL"
            105 -> "LMUL"
            106 -> "FMUL"
            107 -> "DMUL"
            108 -> "IDIV"
            109 -> "LDIV"
            110 -> "FDIV"
            111 -> "DDIV"
            112 -> "IREM"
            113 -> "LREM"
            114 -> "FREM"
            115 -> "DREM"
            116 -> "INEG"
            117 -> "LNEG"
            118 -> "FNEG"
            119 -> "DNEG"
            120 -> "ISHL"
            121 -> "LSHL"
            122 -> "ISHR"
            123 -> "LSHR"
            124 -> "IUSHR"
            125 -> "LUSHR"
            126 -> "IAND"
            127 -> "LAND"
            128 -> "IOR"
            129 -> "LOR"
            130 -> "IXOR"
            131 -> "LXOR"
            132 -> "IINC"
            133 -> "I2L"
            134 -> "I2F"
            135 -> "I2D"
            136 -> "L2I"
            137 -> "L2F"
            138 -> "L2D"
            139 -> "F2I"
            140 -> "F2L"
            141 -> "F2D"
            142 -> "D2I"
            143 -> "D2L"
            144 -> "D2F"
            145 -> "I2B"
            146 -> "I2C"
            147 -> "I2S"
            148 -> "LCMP"
            149 -> "FCMPL"
            150 -> "FCMPG"
            151 -> "DCMPL"
            152 -> "DCMPG"
            153 -> "IFEQ"
            154 -> "IFNE"
            155 -> "IFLT"
            156 -> "IFGE"
            157 -> "IFGT"
            158 -> "IFLE"
            159 -> "IF_ICMPEQ"
            160 -> "IF_ICMPNE"
            161 -> "IF_ICMPLT"
            162 -> "IF_ICMPGE"
            163 -> "IF_ICMPGT"
            164 -> "IF_ICMPLE"
            165 -> "IF_ACMPEQ"
            166 -> "IF_ACMPNE"
            167 -> "GOTO"
            168 -> "JSR"
            169 -> "RET"
            170 -> "TABLESWITCH"
            171 -> "LOOKUPSWITCH"
            172 -> "IRETURN"
            173 -> "LRETURN"
            174 -> "FRETURN"
            175 -> "DRETURN"
            176 -> "ARETURN"
            177 -> "RETURN"
            178 -> "GETSTATIC "
            179 -> "PUTSTATIC"
            180 -> "GETFIELD"
            181 -> "PUTFIELD"
            182 -> "INVOKEVIRTUAL "
            183 -> "INVOKESPECIAL"
            184 -> "INVOKESTATIC"
            185 -> "INVOKEINTERFACE"
            186 -> "INVOKEDYNAMIC"
            187 -> "NEW"
            188 -> "NEWARRAY"
            189 -> "ANEWARRAY"
            190 -> "ARRAYLENGTH"
            191 -> "ATHROW"
            192 -> "CHECKCAST"
            193 -> "INSTANCEOF"
            194 -> "MONITORENTER"
            195 -> "MONITOREXIT"
            197 -> "MULTIANEWARRAY"
            198 -> "IFNULL"
            199 -> "IFNONNULL"
            else -> throw IllegalArgumentException("Invalid opcode")
        }
    }

    @JvmStatic
    fun isReturn(opcode: Int): Boolean {
        when (opcode) {
            Opcodes.IRETURN, Opcodes.LRETURN, Opcodes.FRETURN, Opcodes.DRETURN, Opcodes.ARETURN, Opcodes.RETURN -> return true
        }
        return false
    }

    @JvmStatic
    fun isLocalsStore(opcode: Int): Boolean {
        when (opcode) {
            Opcodes.ASTORE, Opcodes.DSTORE, Opcodes.FSTORE, Opcodes.ISTORE, Opcodes.LSTORE -> return true
        }
        return false
    }

    @TestOnly
    fun toNodes(instruction: Instruction): List<AbstractInsnNode> {
        return toNodes(instruction.insn)
    }

    @TestOnly
    fun toNodes(instruction: InsnList): List<AbstractInsnNode> {
        val nodes = ArrayList<AbstractInsnNode>()
        instruction.forEach(Consumer { e: AbstractInsnNode -> nodes.add(e) })
        return nodes
    }
}